File: scraper.js
const puppeteer = require('puppeteer');
const fs = require('fs').promises;
const createCsvWriter = require('csv-writer').createObjectCsvWriter;

async function scrapeElectionData() {
  const browser = await puppeteer.launch({
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });
  
  try {
    const page = await browser.newPage();
    await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36');
    
    console.log('Navegando a la página...');
    await page.goto('https://eleccionesdepartamentales2025.corteelectoral.gub.uy/ResultadosDepartamentales.htm#', {
      waitUntil: 'networkidle2',
      timeout: 30000
    });
    
    // Esperar a que la página cargue
    await page.waitForSelector('#resultadosDepartamental', { timeout: 10000 });
    
    console.log('Seleccionando Lavalleja...');
    // Ejecutar selectDepto para Lavalleja
    await page.evaluate(() => {
      selectDepto('LAVALLEJA', 'Lavalleja');
    });
    
    // Esperar a que se actualicen los datos
    await page.waitForTimeout(3000);
    
    // Extraer datos de la tabla
    console.log('Extrayendo datos...');
    const data = await page.evaluate(() => {
      const table = document.querySelector('#resultadosDepartamental');
      if (!table) return null;
      
      const result = {
        timestamp: new Date().toISOString()
      };
      
      const rows = table.querySelectorAll('tr');
      
      for (const row of rows) {
        const cols = row.querySelectorAll('div');
        if (cols.length >= 2) {
          const label = cols[0].textContent.trim();
          const valueSpan = cols[1].querySelector('span');
          
          if (valueSpan) {
            const value = valueSpan.textContent.trim();
            
            if (label.includes('Resultados al:')) {
              result.ultimaActualizacion = value;
            } else if (label.includes('Circuitos escrutados:')) {
              result.circuitosEscrutados = value;
            } else if (label.includes('Total de circuitos:')) {
              result.totalCircuitos = parseInt(value);
            } else if (label.includes('Circuitos con observaciones:')) {
              result.circuitosConObservaciones = parseInt(value);
            } else if (label.includes('Total de habilitados:')) {
              result.totalHabilitados = parseInt(value.replace(/[^\d]/g, ''));
            }
          }
        }
      }
      
      return result;
    });
    
    // Extraer datos de partidos
    const partidos = await page.evaluate(() => {
      const partidosData = {};
      const partidoElements = document.querySelectorAll('.row.row-xsm.manito');
      
      for (const elemento of partidoElements) {
        const nombreElement = elemento.querySelector('.lema, .lema-sm');
        const votosElement = elemento.querySelector('.subtotal, .subtotal-sm');
        
        if (nombreElement && votosElement) {
          const nombre = nombreElement.textContent.trim();
          const votos = parseInt(votosElement.textContent.trim().replace(/[^\d]/g, ''));
          // Limpiar nombre del partido para usar como clave
          const nombreLimpio = nombre.replace(/\s+/g, '_').replace(/[^\w]/g, '');
          partidosData[`votos_${nombreLimpio}`] = votos;
        }
      }
      
      return partidosData;
    });
    
    // Combinar todos los datos
    const resultado = {
      ...data,
      ...partidos
    };
    
    return resultado;
    
  } catch (error) {
    console.error('Error durante el scraping:', error);
    throw error;
  } finally {
    await browser.close();
  }
}

async function saveToCSV(data) {
  const filename = 'elecciones_lavalleja.csv';
  
  // Verificar si el archivo existe
  let fileExists = false;
  try {
    await fs.access(filename);
    fileExists = true;
  } catch (error) {
    // Archivo no existe
  }
  
  // Crear headers dinámicamente basado en los datos
  const headers = Object.keys(data).map(key => ({
    id: key,
    title: key
  }));
  
  const csvWriter = createCsvWriter({
    path: filename,
    header: headers,
    append: fileExists
  });
  
  await csvWriter.writeRecords([data]);
  console.log(`Datos guardados en ${filename}`);
}

async function main() {
  console.log('🗳️  Iniciando scraping para Lavalleja...');
  console.log(`⏰ ${new Date().toLocaleString()}`);
  
  try {
    const resultado = await scrapeElectionData();
    await saveToCSV(resultado);
    
    console.log('✅ Scraping completado exitosamente');
    console.log('📊 Datos extraídos:', resultado);
    
    process.exit(0);
  } catch (error) {
    console.error('❌ Error:', error);
    process.exit(1);
  }
}

main();

File: scraper.py
import time
import pandas as pd
from datetime import datetime
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.chrome.service import Service
import os

def setup_driver():
    """Configurar Chrome driver para el scraping"""
    chrome_options = Options()
    chrome_options.add_argument('--headless')
    chrome_options.add_argument('--no-sandbox')
    chrome_options.add_argument('--disable-dev-shm-usage')
    chrome_options.add_argument('--disable-gpu')
    chrome_options.add_argument('--remote-debugging-port=9222')
    
    service = Service(ChromeDriverManager().install())
    return webdriver.Chrome(service=service, options=chrome_options)

def scrape_election_data():
    """Hacer scraping de los datos de elecciones"""
    driver = setup_driver()
    
    try:
        print("🗳️  Navegando a la página...")
        driver.get("https://eleccionesdepartamentales2025.corteelectoral.gub.uy/ResultadosDepartamentales.htm#")
        
        # Esperar a que la página cargue
        wait = WebDriverWait(driver, 10)
        wait.until(EC.presence_of_element_located((By.ID, "resultadosDepartamental")))
        
        print("📍 Seleccionando Lavalleja...")
        # Ejecutar JavaScript para seleccionar Lavalleja
        driver.execute_script("selectDepto('LAVALLEJA', 'Lavalleja');")
        
        # Esperar a que se actualicen los datos
        time.sleep(3)
        
        print("📊 Extrayendo datos...")
        
        # Crear diccionario para los datos
        data = {
            'timestamp': datetime.now().isoformat()
        }
        
        # Extraer datos de la tabla
        table = driver.find_element(By.ID, "resultadosDepartamental")
        rows = table.find_elements(By.TAG_NAME, "tr")
        
        for row in rows:
            divs = row.find_elements(By.TAG_NAME, "div")
            if len(divs) >= 2:
                label = divs[0].text.strip()
                
                # Buscar el span dentro del segundo div
                spans = divs[1].find_elements(By.TAG_NAME, "span")
                if spans:
                    value = spans[0].text.strip()
                    
                    if "Resultados al:" in label:
                        data['ultimaActualizacion'] = value
                    elif "Circuitos escrutados:" in label:
                        data['circuitosEscrutados'] = value
                    elif "Total de circuitos:" in label:
                        data['totalCircuitos'] = int(value) if value.isdigit() else value
                    elif "Circuitos con observaciones:" in label:
                        data['circuitosConObservaciones'] = int(value) if value.isdigit() else value
                    elif "Total de habilitados:" in label:
                        # Limpiar número (quitar puntos, comas)
                        clean_value = ''.join(filter(str.isdigit, value))
                        data['totalHabilitados'] = int(clean_value) if clean_value else 0
        
        # Extraer datos de partidos
        print("🎯 Extrayendo votos por partido...")
        partidos = driver.find_elements(By.CSS_SELECTOR, ".row.row-xsm.manito")
        
        for partido in partidos:
            # Buscar el nombre del partido
            nombre_elements = partido.find_elements(By.CSS_SELECTOR, ".lema, .lema-sm")
            votos_elements = partido.find_elements(By.CSS_SELECTOR, ".subtotal, .subtotal-sm")
            
            if nombre_elements and votos_elements:
                nombre = nombre_elements[0].text.strip()
                votos_text = votos_elements[0].text.strip()
                
                # Limpiar los votos (quitar puntos, comas)
                clean_votos = ''.join(filter(str.isdigit, votos_text))
                votos = int(clean_votos) if clean_votos else 0
                
                # Limpiar nombre para usar como clave
                nombre_limpio = nombre.replace(' ', '_').replace('ñ', 'n')
                # Quitar caracteres especiales pero mantener letras acentuadas básicas
                nombre_limpio = ''.join(c for c in nombre_limpio if c.isalnum() or c == '_')
                
                data[f'votos_{nombre_limpio}'] = votos
        
        return data
        
    except Exception as e:
        print(f"❌ Error durante el scraping: {e}")
        raise
        
    finally:
        driver.quit()

def save_to_csv(data):
    """Guardar los datos en CSV"""
    filename = 'elecciones_lavalleja.csv'
    
    # Convertir a DataFrame
    df_new = pd.DataFrame([data])
    
    # Verificar si el archivo existe
    if os.path.exists(filename):
        # Leer datos existentes y concatenar
        df_existing = pd.read_csv(filename)
        df_combined = pd.concat([df_existing, df_new], ignore_index=True)
    else:
        df_combined = df_new
    
    # Guardar CSV
    df_combined.to_csv(filename, index=False)
    print(f"✅ Datos guardados en {filename}")
    
    # Mostrar últimos datos guardados
    print("📋 Último registro:")
    print(df_new.to_string(index=False))

def main():
    print("🗳️  Iniciando scraping para Lavalleja...")
    print(f"⏰ {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    
    try:
        resultado = scrape_election_data()
        save_to_csv(resultado)
        print("✅ Scraping completado exitosamente")
        
    except Exception as e:
        print(f"❌ Error: {e}")
        exit(1)

if __name__ == "__main__":
    main()